arangod/Graph/Providers/ClusterProvider.cpp

```diff
@@ -0,0 +1,334 @@
+                                  ::vertexUrl + StringUtils::itoa(engine.second),
+                                  leased->bufferRef(), reqOpts));
+  }
+
+
+  for (Future<network::Response>& f : futures) {
+    network::Response const& r = f.get();
+
+    if (r.fail()) {
+      THROW_ARANGO_EXCEPTION(network::fuerteToArangoErrorCode(r));
+    }
+
+    auto payload = r.response().stealPayload();
+    bool isPayloadCached = false;
+
+    VPackSlice resSlice(payload->data());
+    if (!resSlice.isObject()) {
+      // Response has invalid format
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_HTTP_CORRUPTED_JSON);
+    }
+    if (r.statusCode() != fuerte::StatusOK) {
+      // We have an error case here. Throw it.
+      THROW_ARANGO_EXCEPTION(network::resultFromBody(resSlice, TRI_ERROR_INTERNAL));
+    }
+
+    for (auto pair : VPackObjectIterator(resSlice, /*sequential*/ true)) {
+      VertexType vertexKey(pair.key);
+
+      if (!isPayloadCached) {
+        _opts.getCache()->datalake().add(payload);
+        isPayloadCached = true;
+      }
+
+      if (!_opts.getCache()->isVertexCached(vertexKey)) {
+        // Protected by datalake (Cache)
+        _opts.getCache()->cacheVertex(std::move(vertexKey), pair.value);
+      }
+    }
+
+    /* TODO: Needs to be taken care of as soon as we enable shortest paths for ClusterProvider
+    bool forShortestPath = true;
+    if (!forShortestPath) {
+      // Fill everything we did not find with NULL
+      for (auto const& v : vertexIds) {
+        result.try_emplace(v, VPackSlice::nullSlice());
+      }
+    vertexIds.clear();
+    }
+    */
+  }
+
+  // put back all looseEnds we we're able to cache
+  for (auto& lE : looseEnds) {
+    if (!_opts.getCache()->isVertexCached(lE->getVertexIdentifier())) {
+      // if we end up here, we we're not able to cache the requested vertex (e.g. it does not exist)
+      _query->warnings().registerWarning(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND, lE->getVertexIdentifier().toString());
+      _opts.getCache()->cacheVertex(std::move(lE->getVertexIdentifier()), VPackSlice::nullSlice());
+    }
+    result.emplace_back(std::move(lE));
+  }
+}
+
+Result ClusterProvider::fetchEdgesFromEngines(VertexType const& vertex) {
+  // Currently unused. This is just here as a placeholder. Needs to be implemented in the future
+  // as soon as we enable the ClusterProvider globally.
+  auto const* engines = _opts.getCache()->engines();
+
+  // TRI_ASSERT(vertexId.isString() || vertexId.isArray());
+
+  transaction::BuilderLeaser leased(trx());
+  leased->openObject(true);
+  leased->add("backward", VPackValue(_opts.isBackward()));
+  leased->add("keys", VPackValue(vertex.toString()));
+  leased->close();
+
+  auto* pool = trx()->vocbase().server().getFeature<NetworkFeature>().pool();
+
+  network::RequestOptions reqOpts;
+  reqOpts.database = trx()->vocbase().name();
+  reqOpts.skipScheduler = true;  // hack to avoid scheduler queue
+
+  std::vector<Future<network::Response>> futures;
```

Comment:
We need to add a TRI_DEFER to handle all open futures if we throw below.
Otherwise we may end up with unhandled exceptions